ui/src/data-products/element-list/element-list-store.ts

```diff
@@ -0,0 +1,199 @@
+    committedFields: (keyof PlanGeometrySearchState)[];
+};
+
+enum MissingSection {
+    MISSING_SECTION = 'MISSING_SECTION',
+}
+
+export type GeometryTypeIncludingMissing = GeometryType | MissingSection;
+
+export const initialPlanGeometrySearchState: PlanGeometrySearchState = {
+    plan: undefined,
+    searchGeometries: {
+        searchLines: true,
+        searchCurves: true,
+        searchClothoids: true,
+        searchMissingGeometry: false,
+    },
+
+    validationErrors: [],
+    committedFields: [],
+};
+
+export type ContinuousSearchParameters = {
+    startTrackMeter: string;
+    endTrackMeter: string;
+    searchGeometries: SearchGeometries;
+};
+
+export type ElementListContinuousGeometrySearchState = {
+    searchFields: ContinuousSearchParameters;
+    searchParameters: ContinuousSearchParameters;
+
+    validationErrors: ValidationError<ContinuousSearchParameters>[];
+    committedFields: (keyof ContinuousSearchParameters)[];
+};
+
+export const initialContinuousSearchState: ElementListContinuousGeometrySearchState = {
+    searchFields: {
+        startTrackMeter: '',
+        endTrackMeter: '',
+        searchGeometries: {
+            searchLines: true,
+            searchCurves: true,
+            searchClothoids: true,
+            searchMissingGeometry: true,
+        },
+    },
+
+    searchParameters: {
+        startTrackMeter: '',
+        endTrackMeter: '',
+        searchGeometries: {
+            searchLines: true,
+            searchMissingGeometry: true,
+            searchClothoids: true,
+            searchCurves: true,
+        },
+    },
+
+    validationErrors: [],
+    committedFields: [],
+};
+
+export const selectedElementTypes = (
+    searchGeometry: SearchGeometries,
+): GeometryTypeIncludingMissing[] =>
+    [
+        searchGeometry.searchLines ? GeometryType.LINE : undefined,
+        searchGeometry.searchCurves ? GeometryType.CURVE : undefined,
+        searchGeometry.searchClothoids ? GeometryType.CLOTHOID : undefined,
+        searchGeometry.searchMissingGeometry ? MissingSection.MISSING_SECTION : undefined,
+    ].filter(filterNotEmpty);
+
+export const validTrackMeterOrUndefined = (trackMeterCandidate: string) => {
+    if (trackMeterIsValid(trackMeterCandidate)) return trackMeterCandidate;
+    else return undefined;
+};
+
+const TRACK_METER_REGEX = /([0-9]{1,4})\+([0-9]{4})$/;
+
+const hasAtLeastOneTypeSelected = ({
+    searchLines,
+    searchCurves,
+    searchClothoids,
+    searchMissingGeometry,
+}: SearchGeometries) => searchLines || searchCurves || searchClothoids || searchMissingGeometry;
+
+export const trackMeterIsValid = (trackMeter: string) => TRACK_METER_REGEX.test(trackMeter);
+
+const validateContinuousGeometry = (
+    state: ElementListContinuousGeometrySearchState,
+): ValidationError<ContinuousSearchParameters>[] =>
+    [
+        hasAtLeastOneTypeSelected(state.searchFields.searchGeometries)
+            ? undefined
```

Comment:
Tämä ternääri rakenne toistuu aika usein. Mietin olisiko js:ssäkin nätimpää tehdä kotlinmainen validate-funkkari jota voisi kutsua näin:
```
validate(hasAtLeastOneTypeSelected(...), () => { 
  field: ... 
}),
```
Eli käytännössä se poistaa näistä vain tuon ternäärin `? undefined :` osan ja korvaa sen funktiokutsulla, mutta kutsulla on nimi -> hiukan kyseenalainen. Toki olisi myös mahdollista antaa lambdan tilalle nuo errorin kentät parametreina, jolloin `as keyof` voi hoitaa parametrin tyypillä ja tyypille voi antaa default-arvona ERROR, ja kutsupaikka näyttäisi enää tältä:
```
validate(hasAtLeastOneTypeSelected(...), 'searchGeometries', 'no-types-selected'),
```

Reply:
Tuo nykyinen validointikoodi oli niin rumaa, että refaktoroin tuota vähän. Nyt on validate-funktio